python/lsst/cell_coadds/_grid_container.py

```diff
@@ -0,0 +1,201 @@
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+from __future__ import annotations
+
+from collections.abc import MutableMapping
+from itertools import product
+from typing import TYPE_CHECKING, Callable, Iterable, Iterator, TypeVar
+
+from lsst.skymap import Index2D
+
+from ._uniform_grid import UniformGrid
+
+if TYPE_CHECKING:
+    import lsst.geom as geom
+
+
+__all__ = ("GridContainer",)
+
+T = TypeVar("T")
+
+
+class GridContainer(MutableMapping[Index2D, T]):  # Generic, Collection
+    """A container whose elements form a 2-d grid."""
+
+    def __init__(self, shape: Index2D, offset: Index2D | None = None) -> None:
+        super().__init__()
+
+        self._offset = offset if offset else Index2D(0, 0)
+        self._shape = shape
+        self._cells: dict[Index2D, T] = {}
+        self._mapping = self._cells
+
+    def _check(self, index: Index2D) -> None:
+        """Check if a given index belongs to the container or not."""
+        if index.x < self.offset.x or index.x >= self.offset.x + self.shape.x:
+            raise ValueError(
+                f"x index {index.x} out of range; expected a value between {self.offset.x} and "
+                f"{self.offset.x + self.shape.x -1}."
+            )
+        if index.y < self.offset.y or index.y >= self.offset.y + self.shape.y:
+            raise ValueError(
+                f"y index {index.y} out of range; expected a value between {self.offset.y} and "
+                f"{self.offset.y + self.shape.y -1}."
+            )
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}(shape={self.shape}, offset={self.offset})"
+
+    # Implement the necessary methods for MutableMapping.
+    def __len__(self) -> int:
+        return len(self._cells)
+
+    def __getitem__(self, index: Index2D) -> T:
+        return self._cells[index]
+
+    # def __contains__(self, index: object) -> bool:
+    #     return index in self._cells
+
+    def __setitem__(self, index: Index2D, value: T) -> None:
+        self._check(index)
+        self._cells[index] = value
+
+    def __delitem__(self, index: Index2D) -> None:
+        self._check(index)
+        del self._cells[index]
+
+    def __iter__(self) -> Iterator[T]:
+        yield from self._cells.values()
+
+    def keys(self) -> Iterable[Index2D]:  # populated_indices
```

Comment:
Because it is a `MutableMapping` now, `keys` method is expected and I think what this does is a very reasonable thing here. The docstring references to `indices` as well.

Reply:
After `__iter__` is switched to iterate over keys, you can drop this method entirely as `MutableMapping` will provide one for you.  It's also important that it return a `KeysView`, which is set-like, not just an iterator.